R/hiper_glm.R

```diff
@@ -1,31 +1,104 @@
+
+####################### Helper functions ######################
+
+########## For cholesky ########
+## Find linear regression coefficients estimates via MLE,
+## with cholesky decomposition while solving matrices
+chol_finder <- function(design, outcome){
+  ## With least-square, we have,
+  ## beta_est = (X^T X) X^T y.
+  ## Which is solving beta_est in (X^T X) beta_est = X^T y.
+  ## This is equivalent to solving Ax = b,
+  ## where, A and b are the following
+
+  A <- t(design) %*% design
+  b <- t(design) %*% outcome
+
+  ## solve "Ax = b" question via cholesky decomp
+  L <- chol(A)
+  beta_est <- backsolve(t(L), forwardsolve(L, t(A) %*% b))
+  return(c(beta_est))
+}
+
+
+############ FOR BFGS ############
+## The log-likelihood function
+gaussian_logLi <- function(beta, X, Y, noise_var = 1){
+  sigma <- sqrt(noise_var)
+  Sigma_inv <- (1/sigma) * diag(nrow = length(Y))
+
+  f <- -0.5 * t(Y - X %*% beta) %*% Sigma_inv %*% (Y - X %*% beta)
+  return(f)
+}
+## The gradient
+gradient <- function(beta, X = design, Y = outcome){
+  quad_grad <- function(x, Sigma_inv) {
+    return(-Sigma_inv %*% x)
+  }
+
+  gr = - t(X) %*% quad_grad(x = Y - X %*% beta,
+                                Sigma_inv = diag(nrow = length(Y)))
+  return(gr)
+}
+
+##
+BFGS_finder <- function(design, outcome){
+  beta_init <- c(0,0,0,0)
```

Comment:
-1

What if the design matrix has more than 4 columns? Sure, R might recycle the vector (I haven't checked), but it's confusing as written.